File: src/main/java/ua/softgroup/matrix/server/persistent/repository/WorkDayRepository.java
package ua.softgroup.matrix.server.persistent.repository;

import org.springframework.data.jpa.repository.Query;
import org.springframework.data.repository.CrudRepository;
import org.springframework.data.repository.query.Param;
import ua.softgroup.matrix.server.persistent.entity.Project;
import ua.softgroup.matrix.server.persistent.entity.User;
import ua.softgroup.matrix.server.persistent.entity.WorkDay;

import java.time.LocalDate;
import java.util.Set;

public interface WorkDayRepository extends CrudRepository<WorkDay, Long> {

    Set<WorkDay> findByAuthorAndProject(User author, Project project);

    Set<WorkDay> findByDate(LocalDate date);

    @Query(value = "SELECT sum(work_seconds) FROM work_day WHERE author_id = :userId and date = :date", nativeQuery = true)
    Integer getTotalWorkSeconds(@Param("userId") Long userId, @Param("date") LocalDate date);

    @Query(value = "SELECT sum(idle_seconds) FROM work_day WHERE author_id = :userId and date = :date", nativeQuery = true)
    Integer getTotalIdleSeconds(@Param("userId") Long userId, @Param("date") LocalDate date);

    Set<WorkDay> findByAuthorAndDate(User author, LocalDate date);

    Set<WorkDay> findByProjectSupervisorIdAndDate(Long projectSupervisorId, LocalDate date);

    Set<WorkDay> findByProjectSupervisorIdAndDateBetween(Long supervisorId, LocalDate from, LocalDate to);

    Set<WorkDay> findByAuthorIdAndProjectSupervisorIdAndDateBetween(Long authorId, Long supervisorId, LocalDate from, LocalDate to);

    Set<WorkDay> findByAuthorIdAndDateBetween(Long authorId, LocalDate from, LocalDate to);

    Set<WorkDay> findByAuthorIdAndCheckedFalse(Long authorId);

    Set<WorkDay> findByProjectSupervisorIdAndCheckedFalse(Long supervisorId);

    Set<WorkDay> findByCheckedFalse();

    Set<WorkDay> findByDateBetween(LocalDate from, LocalDate to);

    WorkDay findByAuthorAndProjectAndDate(User author, Project project, LocalDate localDate);

    @Query(value = "SELECT sum(work_seconds) FROM work_day WHERE project_id = :projectId and author_id = :userId", nativeQuery = true)
    Integer getTotalWorkSeconds(@Param("userId") Long userId, @Param("projectId") Long projectId);

    @Query(value = "SELECT sum(idle_seconds) FROM work_day "
                                          + "WHERE project_id = :projectId "
                                                 + "AND author_id = :userId "
                                                 + "AND date BETWEEN :startDate AND :endDate", nativeQuery = true)
    Integer getCurrentMonthIdleSeconds(@Param("userId") Long userId, @Param("projectId") Long projectId,
                                       @Param("startDate") LocalDate startDate, @Param("endDate") LocalDate endDate);

    @Query(value = "SELECT sum(idle_seconds) FROM work_day WHERE project_id = :projectId AND author_id = :userId", nativeQuery = true)
    Integer getTotalIdleSeconds(@Param("userId") Long userId, @Param("projectId") Long projectId);

    @Query(value = "SELECT sum(work_seconds) FROM work_day "
                 + "WHERE author_id = :userId AND date BETWEEN :startDate AND :endDate", nativeQuery = true)
    Integer getTotalWorkSeconds(@Param("userId") Long userId, @Param("startDate") LocalDate startDate, @Param("endDate") LocalDate endDate);

    @Query(value = "SELECT sum(idle_seconds) FROM work_day "
                 + "WHERE author_id = :userId AND date BETWEEN :startDate AND :endDate", nativeQuery = true)
    Integer getTotalIdleSeconds(@Param("userId") Long userId, @Param("startDate") LocalDate startDate, @Param("endDate") LocalDate endDate);

    @Query(value = "SELECT sum(symbols_count) FROM work_day "
                 + "WHERE author_id = :userId AND date BETWEEN :startDate AND :endDate", nativeQuery = true)
    Integer getSymbolsCount(@Param("userId") Long userId, @Param("startDate") LocalDate startDate, @Param("endDate") LocalDate endDate);

    @Query(value = "SELECT sum(windows_switched_count) FROM work_day "
                 + "WHERE author_id = :userId AND date BETWEEN :startDate AND :endDate", nativeQuery = true)
    Integer getWindowsSwitchedCount(@Param("userId") Long userId, @Param("startDate") LocalDate startDate, @Param("endDate") LocalDate endDate);

}

File: src/main/scala/ua/softgroup/matrix/server/persistent/repository/WorkDayRepository.scala
package ua.softgroup.matrix.server.persistent.repository

import java.time.LocalDate
import java.util
import java.util.Set

import org.springframework.data.jpa.repository.Query
import org.springframework.data.repository.CrudRepository
import org.springframework.data.repository.query.Param
import ua.softgroup.matrix.server.persistent.entity.{Project, User, WorkDay}

/**
  * @author Oleksandr Tyshkovets <sg.olexander@gmail.com> 
  */
trait WorkDayRepository extends CrudRepository[WorkDay, java.lang.Long] {

  def findByAuthorAndProject(author: User, project: Project): util.Set[WorkDay]

  def findByDate(date: LocalDate): util.Set[WorkDay]

  @Query(value = "SELECT sum(work_seconds) FROM work_day " +
                 "WHERE author_id = :userId and date = :date", nativeQuery = true)
  def getTotalWorkSeconds(@Param("userId") userId: Long, @Param("date") date: LocalDate): Integer

  @Query(value = "SELECT sum(idle_seconds) FROM work_day " +
                 "WHERE author_id = :userId and date = :date", nativeQuery = true)
  def getTotalIdleSeconds(@Param("userId") userId: Long, @Param("date") date: LocalDate): Integer

  def findByAuthorAndDate(author: User, date: LocalDate): util.Set[WorkDay]

  def findByProjectSupervisorIdAndDate(projectSupervisorId: Long, date: LocalDate): util.Set[WorkDay]

  def findByProjectSupervisorIdAndDateBetween(supervisorId: Long, from: LocalDate, to: LocalDate): util.Set[WorkDay]

  def findByAuthorIdAndProjectSupervisorIdAndDateBetween(authorId: Long, supervisorId: Long, from: LocalDate, to: LocalDate): util.Set[WorkDay]

  def findByAuthorIdAndDateBetween(authorId: Long, from: LocalDate, to: LocalDate): util.Set[WorkDay]

  def findByAuthorIdAndCheckedFalse(authorId: Long): util.Set[WorkDay]

  def findByProjectSupervisorIdAndCheckedFalse(supervisorId: Long): util.Set[WorkDay]

  def findByCheckedFalse: util.Set[WorkDay]

  def findByDateBetween(from: LocalDate, to: LocalDate): util.Set[WorkDay]

  def findByAuthorAndProjectAndDate(author: User, project: Project, localDate: LocalDate): WorkDay

  @Query(value = "SELECT sum(work_seconds) FROM work_day " +
                 "WHERE project_id = :projectId and author_id = :userId", nativeQuery = true)
  def getTotalWorkSeconds(@Param("userId") userId: Long, @Param("projectId") projectId: Long): Integer

  @Query(value = "SELECT sum(idle_seconds) FROM work_day " +
                 "WHERE project_id = :projectId AND author_id = :userId AND date BETWEEN :startDate AND :endDate", nativeQuery = true)
  def getCurrentMonthIdleSeconds(@Param("userId") userId: Long,
                                 @Param("projectId") projectId: Long,
                                 @Param("startDate") startDate: LocalDate,
                                 @Param("endDate") endDate: LocalDate): Integer

  @Query(value = "SELECT sum(idle_seconds) FROM work_day " +
                 "WHERE project_id = :projectId AND author_id = :userId", nativeQuery = true)
  def getTotalIdleSeconds(@Param("userId") userId: Long, @Param("projectId") projectId: Long): Integer

  @Query(value = "SELECT sum(work_seconds) FROM work_day " +
                 "WHERE author_id = :userId AND date BETWEEN :startDate AND :endDate", nativeQuery = true)
  def getTotalWorkSeconds(@Param("userId") userId: Long,
                          @Param("startDate") startDate: LocalDate,
                          @Param("endDate") endDate: LocalDate): Integer

  @Query(value = "SELECT sum(idle_seconds) FROM work_day " +
                 "WHERE author_id = :userId AND date BETWEEN :startDate AND :endDate", nativeQuery = true)
  def getTotalIdleSeconds(@Param("userId") userId: Long,
                          @Param("startDate") startDate: LocalDate,
                          @Param("endDate") endDate: LocalDate): Integer

  @Query(value = "SELECT sum(symbols_count) FROM work_day " +
                 "WHERE author_id = :userId AND date BETWEEN :startDate AND :endDate", nativeQuery = true)
  def getSymbolsCount(@Param("userId") userId: Long,
                      @Param("startDate") startDate: LocalDate,
                      @Param("endDate") endDate: LocalDate): Integer

  @Query(value = "SELECT sum(windows_switched_count) FROM work_day " +
                 "WHERE author_id = :userId AND date BETWEEN :startDate AND :endDate", nativeQuery = true)
  def getWindowsSwitchedCount(@Param("userId") userId: Long,
                              @Param("startDate") startDate: LocalDate,
                              @Param("endDate") endDate: LocalDate): Integer

}
